Update to payment doc for Janitor workflow and other minor changes (#435)
killbill/killbill-docs: userguide/payment/includes/payment-states.adoc

If a payment is aborted by a control plugin, a payment attempt row is recorded with a state of ABORTED (no payment nor payment transaction is created). Kill Bill will return a 422 HTTP code.

[[Janitor]]
=== Janitor

> PENDING and UNKNOWN transaction states should be temporary: the payment plugin needs to eventually tell Kill Bill what the final state of these transactions should be. This can be done explicitly by the plugin by calling the https://github.com/killbill/killbill-api/blob/4ae1c343a593de937415e21feecb9f5405037fa3/src/main/java/org/killbill/billing/payment/api/PaymentApi.java#L402[`PaymentApi.notifyPendingTransactionOfStateChanged`] (for instance, when receiving a webhook event from the gateway) or implicitly by the Janitor.

Reviewer: ```suggestion
PENDING and UNKNOWN transaction states should be temporary; the payment plugin needs to eventually tell Kill Bill what the final state of these transactions should be. This can be done explicitly by the plugin by calling the https://github.com/killbill/killbill-api/blob/4ae1c343a593de937415e21feecb9f5405037fa3/src/main/java/org/killbill/billing/payment/api/PaymentApi.java#L402[`PaymentApi.notifyPendingTransactionOfStateChanged`] (for instance, when receiving a webhook event from the gateway) or implicitly by the Janitor).
```

> PENDING and UNKNOWN transaction states should be temporary: the payment plugin needs to eventually tell Kill Bill what the final state of these transactions should be. This can be done explicitly by the plugin by calling the https://github.com/killbill/killbill-api/blob/4ae1c343a593de937415e21feecb9f5405037fa3/src/main/java/org/killbill/billing/payment/api/PaymentApi.java#L402[`PaymentApi.notifyPendingTransactionOfStateChanged`] (for instance, when receiving a webhook event from the gateway) or implicitly by the Janitor.

Reviewer: Use a semicolon instead of colon to separate two sentences that can each stand on their own. 
```suggestion
PENDING and UNKNOWN transaction states should be temporary; the payment plugin needs to eventually tell Kill Bill what the final state of these transactions should be. This can be done explicitly by the plugin by calling the https://github.com/killbill/killbill-api/blob/4ae1c343a593de937415e21feecb9f5405037fa3/src/main/java/org/killbill/billing/payment/api/PaymentApi.java#L402[`PaymentApi.notifyPendingTransactionOfStateChanged`] (for instance, when receiving a webhook event from the gateway) or implicitly by the Janitor.
```

Author: got it 👍

Reviewer: Cool!

> The Janitor refers to a set of internal Kill Bill features to help ensure transactions end up in a terminal state: these are the Janitor notification queue and the on-the-fly Janitor. They both work by calling the plugin (via https://github.com/killbill/killbill-plugin-api/blob/d9eca5af0e37541069b1c608f95e100dbe13b301/payment/src/main/java/org/killbill/billing/payment/plugin/api/PaymentPluginApi.java#L144[`PaymentPluginApi.getPaymentInfo`]) and updating the state owned by Kill Bill if it has changed. It is imperative payment plugin developers take great care when implementing this API for this mechanism to work (see our Payment Plugin developer documentation for more details).

Reviewer: ```suggestion
The Janitor refers to a set of internal Kill Bill features to help ensure transactions end up in a terminal state; these are the Janitor notification queue and the on-the-fly Janitor. They both work by calling the plugin (via https://github.com/killbill/killbill-plugin-api/blob/d9eca5af0e37541069b1c608f95e100dbe13b301/payment/src/main/java/org/killbill/billing/payment/plugin/api/PaymentPluginApi.java#L144[`PaymentPluginApi.getPaymentInfo`]) and updating the state owned by Kill Bill if it has changed. It is imperative payment plugin developers take great care when implementing this API for this mechanism to work (see our Payment Plugin developer documentation for more details).
```
The semi-colon is correct to use to join two closely related *complete* sentences (each can stand alone). The colon is appropriate to use in front of a short list or, in the above case, "The Janitor refers to a set of internal Kill Bill features to help ensure transactions end up in a terminal state: the Janitor notification queue and the on-the-fly Janitor." (I removed "these are.")

> The Janitor refers to a set of internal Kill Bill features to help ensure transactions end up in a terminal state: these are the Janitor notification queue and the on-the-fly Janitor. They both work by calling the plugin (via https://github.com/killbill/killbill-plugin-api/blob/d9eca5af0e37541069b1c608f95e100dbe13b301/payment/src/main/java/org/killbill/billing/payment/plugin/api/PaymentPluginApi.java#L144[`PaymentPluginApi.getPaymentInfo`]) and updating the state owned by Kill Bill if it has changed. It is imperative payment plugin developers take great care when implementing this API for this mechanism to work (see our Payment Plugin developer documentation for more details).

Reviewer: ```suggestion
The Janitor refers to a set of internal Kill Bill features to help ensure transactions end up in a terminal state; these are the Janitor notification queue and the on-the-fly Janitor. They both work by calling the plugin (via https://github.com/killbill/killbill-plugin-api/blob/d9eca5af0e37541069b1c608f95e100dbe13b301/payment/src/main/java/org/killbill/billing/payment/plugin/api/PaymentPluginApi.java#L144[`PaymentPluginApi.getPaymentInfo`]) and updating the state owned by Kill Bill if it has changed. For this mechanism to work, it is *imperative* that payment plugin developers take great care when implementing this API (see our https://docs.killbill.io/latest/payment_plugin.html[__Payment Plugin__] developer documentation for more details).
```

> ==== Janitor notification queue
> 
> A Janitor notification entry is inserted whenever a payment results in a non terminal state (whether the payment was initiated internally by Kill Bill or externally via our payment APIs). Kill Bill will then periodically poll the plugin to attempt to fix the state.

Reviewer: ```suggestion
==== Janitor Notification Queue

A Janitor notification entry is inserted whenever a payment results in a non-terminal state (whether the payment was initiated internally by Kill Bill or externally via our payment APIs). Kill Bill then periodically polls the plugin to attempt to fix the state.
```

> The polling delay is configured via the  `org.killbill.payment.janitor.unknown.retries` property for UNKNOWN transactions (default `5m,1h,1d,1d,1d,1d,1d`) and the `org.killbill.payment.janitor.pending.retries` property for PENDING transactions (default `1h, 1d`). These properties can be configured as explained in the https://docs.killbill.io/latest/userguide_configuration.html[_Configuration Guide_]. Companies relying on PENDING for offline payment methods (ACH, Boleto, etc.) should increase the number of attempts, as such payments can be slow (they can take multiple days to execute).

Reviewer: ```suggestion
The polling delay is configured via the  `org.killbill.payment.janitor.unknown.retries` property for UNKNOWN transactions (default `5m,1h,1d,1d,1d,1d,1d`) and the `org.killbill.payment.janitor.pending.retries` property for PENDING transactions (default `1h,1d`). You can configure these properties as explained in the https://docs.killbill.io/latest/userguide_configuration.html[_Configuration Guide_]. Companies that rely on PENDING for offline payment methods (ACH, Boleto, etc.) should increase the number of attempts, as such payments can be slow (multiple days to execute).
```

> The polling delay is configured via the  `org.killbill.payment.janitor.unknown.retries` property for UNKNOWN transactions (default `5m,1h,1d,1d,1d,1d,1d`) and the `org.killbill.payment.janitor.pending.retries` property for PENDING transactions (default `1h, 1d`). These properties can be configured as explained in the https://docs.killbill.io/latest/userguide_configuration.html[_Configuration Guide_]. Companies relying on PENDING for offline payment methods (ACH, Boleto, etc.) should increase the number of attempts, as such payments can be slow (they can take multiple days to execute).

Reviewer: ```suggestion
The polling delay is configured via the `org.killbill.payment.janitor.unknown.retries` property for UNKNOWN transactions (default `5m,1h,1d,1d,1d,1d,1d`) and the `org.killbill.payment.janitor.pending.retries` property for PENDING transactions (default `1h,1d`). These properties can be configured as explained in the https://docs.killbill.io/latest/userguide_configuration.html[_Configuration Guide_]. Companies relying on PENDING for offline payment methods (ACH, Boleto, etc.) should increase the number of attempts, as such payments can be slow (i.e., take multiple days to execute).
```

> ==== On-the-fly Janitor

Reviewer: ```suggestion
==== On-The-Fly Janitor
```
I checked a capitalization tool, and most style guides capitalized the "the" (because it's part of the overall descriptive adjective, I suppose).

The on-the-fly Janitor is invoked:

> * Whenever a payment (or set of payments) is https://github.com/killbill/killbill-api/blob/4ae1c343a593de937415e21feecb9f5405037fa3/src/main/java/org/killbill/billing/payment/api/PaymentApi.java#L439[retrieved via API] and the flag `withPluginInfo` is set. In this case, a GET operation might result in data being written to disk (but it's not really new state, it's just existing state being updated that was incomplete).

Reviewer: ```suggestion
* Whenever a payment (or set of payments) is https://github.com/killbill/killbill-api/blob/4ae1c343a593de937415e21feecb9f5405037fa3/src/main/java/org/killbill/billing/payment/api/PaymentApi.java#L439[retrieved via API] and the flag `withPluginInfo` is set. In this case, a `GET` operation might result in data being written to disk (but it's not really new state, it's just existing state being updated that was incomplete).
```
Not sure about this change.

> * Whenever a payment (or set of payments) is https://github.com/killbill/killbill-api/blob/4ae1c343a593de937415e21feecb9f5405037fa3/src/main/java/org/killbill/billing/payment/api/PaymentApi.java#L439[retrieved via API] and the flag `withPluginInfo` is set. In this case, a GET operation might result in data being written to disk (but it's not really new state, it's just existing state being updated that was incomplete).

Reviewer: ```suggestion
* Whenever a payment (or set of payments) is https://github.com/killbill/killbill-api/blob/4ae1c343a593de937415e21feecb9f5405037fa3/src/main/java/org/killbill/billing/payment/api/PaymentApi.java#L439[retrieved via API] and the flag `withPluginInfo` is set. In this case, a `GET` operation might result in data being written to disk (but it's not really a new state, it's just the existing state being updated that was incomplete).
```
Not sure the `GET` change is right.

* Whenever an operation is performed on an existing payment (either initiated internally by Kill Bill or externally via our payment APIs): the Janitor is invoked first to get the latest state to ensure the internal state machine prevents disallowed transitions in case the state couldn't be fixed (or if it's already in a final state). For example, if an authorization state is currently UNKNOWN and you attempt to capture it, the Janitor will ask the plugin first to fix the state and either let the capture go through (if the authorization is fixed as SUCCESS) or disallow it (if the authorization couldn't be fixed or was unsuccessful).

Note however that the `notifyPendingPaymentOfStateChanged` operation will not invoke the Janitor (as this API is often called from payment plugins, we don't want the state to be fixed twice).

==== Janitor and Control Plugins

> When the Janitor fixes a payment, it will re-run the Payment Control state machine (i.e. go through Payment Control plugins) if needed. So for instance, invoice payments that were created as part of a RECURRING invoice may end up being fixed automatically by the Janitor, and the balance associated with the invoice would then be correctly reflected.

Reviewer: ```suggestion
When the Janitor fixes a payment, it will re-run the Payment Control state machine (i.e. go through Payment Control plugins) if needed. For instance, invoice payments that were created as part of a RECURRING invoice may end up being fixed automatically by the Janitor, and the balance associated with the invoice would then be correctly reflected.
```
